### PR TITLE
feat(sns): Automatically trigger upgrades when target_version is ahead of the current version

### PR DESCRIPTION
**Pull request by anchpop**
This PR adds a new periodic task (or rather, a new thing to be called from our existing periodic task) that will initiate an upgrade when one is necessary.

At the time of writing the proto backwards compatibility checker is complaining about this:

```
ic_sns_governance/pb/v1/governance.proto:1427:5:Field "4" with name "proposal_id" on message "PendingVersion" changed cardinality from "optional with implicit presence" to "optional with explicit presence".
```

Which I don't really think is a real issue, so I'm tagging the PR with CI_OVERRIDE_BUF_BREAKING

[← Previous PR](https://github.com/dfinity/ic/pull/2453)

### CODE REVIEW

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:2635]: Pleas rename this?

File: rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto
Comment (anchpop) [inline rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto:1520]: The pending_version in the governance proto has been changed to have `proposal_id` be optional. This change keeps `GetRunningSnsVersionResponse`'s API the same – if `proposal_id` were changed to be optional in the API, it would no longer be backwards-compatible (as candid cannot decode `T` if it was encoded as `opt T`).

File: rs/nervous_system/integration_tests/tests/advance_target_version.rs
Comment (aterga) [inline rs/nervous_system/integration_tests/tests/advance_target_version.rs:340]: Let's make these pushes to the upgrade journal 1-liners by using some helpers.

File: rs/sns/governance/canister/governance_test.did
Comment (aterga) [inline rs/sns/governance/canister/governance_test.did:269]: Let's not rename the response API type, instead the input API type can be renamed.

File: rs/nervous_system/integration_tests/tests/advance_target_version.rs
Comment (aterga) [reply rs/nervous_system/integration_tests/tests/advance_target_version.rs:340]: This can happen in a follow-up PR.

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:2549]: Let's add a comment to explain why we need two similar-looking functions to co-exist.

Comment (aterga) [inline rs/sns/governance/src/governance.rs:2549]: On second thoughts (and as discussed over zoom), 

let's have just one function that takes `Option<ProposalId>`

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:2569]: Indicate when that lock was acquired?

Or even have a fall back logic for unlocking?

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:2698]: What is this returning?

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:2748]:
```suggestion
        Ok(())
```

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:4846]: Nit 

```suggestion
    async fn initiate_upgrade_if_sns_behind_target_version(&mut self) {
```

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:4853]: In this case, logging is probably worth it, as we're observing an inconsistent state.

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:4857]: Ditto

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:4874]:
```suggestion
            log!(ERROR, "Deployed version {} is not on the upgrade path {}", deployed_version.human_readable(), );
```

```
impl Version {
  pub fn human_readable(&self) -> String {
        format!(
            "SnsVersion {{ root:{}, governance:{}, swap:{}, index:{}, ledger:{}, archive:{} }}",
            hash_to_hex_string(self.root_wasm_hash),
            hash_to_hex_string(self.governance_wasm_hash),
            hash_to_hex_string(self.swap_wasm_hash),
            hash_to_hex_string(self.index_wasm_hash),
            hash_to_hex_string(self.ledger_wasm_hash),
            hash_to_hex_string(self.archive_wasm_hash),
        )
    }
}
```

Or implement `Display for Version`.

Comment (aterga) [reply rs/sns/governance/src/governance.rs:4874]: Instead of logging, this information should in added to the upgrade journal.

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:4881]: Instead of logging, this information should in added to the upgrade journal.

File: rs/sns/governance/src/governance.rs
Comment (aterga) [inline rs/sns/governance/src/governance.rs:4885]: Let's detect separately the case when `target_position < deployed_position`, and reset the target in that case.

This should also produce a journal entry.

File: rs/sns/governance/src/governance.rs
Comment (max-dfinity) [inline rs/sns/governance/src/governance.rs:2438]: I think you need to rebase onto latest after merging the previous PR?

File: rs/sns/governance/src/governance.rs
Comment (max-dfinity) [inline rs/sns/governance/src/governance.rs:4880]: I've seen this twice, and it feels like it should live in a function.

File: rs/sns/governance/src/governance.rs
Comment (max-dfinity) [inline rs/sns/governance/src/governance.rs:2732]: Just so you know, this is going to return before the upgrade is actually finished, which I don't think works correctly with the control flow later on.

File: rs/sns/governance/canister/governance_test.did
Comment (anchpop) [reply rs/sns/governance/canister/governance_test.did:269]: Done!

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs]: Done!

Comment (anchpop) [reply rs/sns/governance/src/governance.rs]: Addressed here https://github.com/dfinity/ic/pull/2034#discussion_r1831410370

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs]: Outdated

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs:4845]: This is actually an expected state, not inconsistent

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs]: Done

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs]: Nothing, but fixed now!

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs]: Done

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs]: Done

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs:4874]: So does that mean that we should reset the target version in this state?

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs:4878]: Holding off on this as it'll be unnecessary after #2553 is merged

File: rs/nervous_system/common/src/lib.rs
Comment (aterga) [inline rs/nervous_system/common/src/lib.rs:786]: Could we use lower case?

**Review by aterga — APPROVED** [review]

File: rs/nervous_system/common/src/lib.rs
Comment (anchpop) [reply rs/nervous_system/common/src/lib.rs]: Done!

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs:4868]: Did you mean the whole thing including the `else {...}`, or just the `upgrade_steps.iter().position(|v| v == &target_version)` part?

File: rs/sns/governance/src/governance.rs
Comment (andrew-lee-work) [inline rs/sns/governance/src/governance.rs:4907]: Is the result of the call to upgrade the canister journaled? I would recommend journaling the result of each action that you journal. It seems here the start of the upgrade is journaled, but not the result.

I see that you journal the `invalidate_target_version`, but that is missing the context that `upgrade_sns_framework_canister` errored out.

File: rs/sns/governance/src/governance.rs
Comment (andrew-lee-work) [inline rs/sns/governance/src/governance.rs:4829]: It seems an upgrade proposal could be adopted and be "in progress" while awaiting the root to upgrade the SNS. Is there a lock preventing this?

**Review by andrew-lee-work — APPROVED** [review]
LGTM, other than the other comments I left.

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs:4829]: `check_no_upgrades_in_progress` is exactly what detects this case and prevents us from initiating an upgrade if necessary

File: rs/nervous_system/integration_tests/tests/advance_target_version.rs
Comment (anchpop) [reply rs/nervous_system/integration_tests/tests/advance_target_version.rs:310]: Done

File: rs/sns/governance/src/governance.rs
Comment (anchpop) [reply rs/sns/governance/src/governance.rs:4923]: I added the context! Journaling the result of the canister call seem excessive to me

File: rs/sns/governance/src/governance.rs
Comment (max-dfinity) [reply rs/sns/governance/src/governance.rs:4868]: Sort of feels like it could be a method on CachedUpgradeSteps to get the position of a given version in the upcoming steps...

**Pull request merged**
